### ISSUE TITLE
Should the debug messages of the well model not be printed to the terminal by default?

### ISSUE THREAD

**Issue by andlaus** [OPM/opm-simulators#1703]
when I run flow on e.g. Norne, the terminal gets cluttered by quite a few messages like this

```
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Switching control mode for well C-1H from BHP to SURFACE_RATE
```

Since they are not interesting for anyone who is not developing on the well model, they IMHO these should only go into the DBG and maybe the PRT file. I had a brief look at fixing this, but the `WellSwitchingLogger` class defeated me...

**Comment by GitPaean** [comment]
> they are not interesting for anyone who is not developing on the well model

I am not sure about this. Some users/engineers might be interested in this. 

Anyway, it can be discussed.

**Comment by andlaus** [comment]
> I am not sure about this. Some users/engineers might be interested in this.

sure, but then they can look at the DBG file, just like for FIP numbers. Also, the terminal output is currently pretty useless because it is impossible to say at which iteration it happened.

**Comment by alfbr** [comment]
> Since they are not interesting for anyone who is not developing on the well model

On what basis do you claim that? Are you working with users who have complained? Not saying that you are wrong, just want context.

**Comment by andlaus** [comment]
> On what basis do you claim that?

I claim this on the basis that personally I'm even less interested in these messages than in the FIP numbers and that these messages tend to be annoying if many wells switch their control mode often. That said, I don't think that this is a big issue, so if you insist that cluttering the terminal with these messages without any context, I'm fine with the current state of affairs. (I just think that the current policy what the user sees on the terminal is very inconsistent because more important information is *not* printed.)

**Comment by GitPaean** [comment]
> very inconsistent because more important information is not printed

What do you mean by `more important information` here? Well control mode switching are actually quite important. I do not insist they should be put on screen.

**Comment by andlaus** [comment]
> What do you mean by more important information here? 

for people like me, this is e.g., convergence information of the newton solver during a time step. (I don't complain, I don't have an issue with looking at the DBG file), for reservoir engineers this is e.g., the FIP data.

**Comment by andlaus** [comment]
> Well control mode switching are actually quite important.

I agree that this is the case for people who are primarily concerned with developing the well model.

> I  do not insist they should be put on screen.

that's my point :)

**Comment by GitPaean** [comment]
Maybe for the well control, for the screen (even PR file), we can just tell the converged well control mode at the end of the time step, not giving the back-and-forth switching process.

**Comment by andlaus** [comment]
> Maybe for the well control, for the screen (even PR file), we can just tell the converged well control mode at the end of the time step, not giving the back-and-forth switching process.

that also would be a good option IMO.

**Comment by OPMUSER** [comment]
Cannot somebody verify my earlier explanation, before changes are made?

**Comment by atgeirr** [comment]
> The impression I get from the Norne model, is that when Flow switches the control mode from SURFACE-RATE to BHP it issues a MESSAGE. However, at the next time step, it appears that still tries to operate the well at SURFACE-RATE, which it cannot, so it then changes control to BHP and issues a message. And I tthink his is why we are getting all the messages.

This used to be true for a while, at least the 2017.10 and 2018.04 releases have this behaviour. However in the current master this is fixed, and you do not get repeated messages about well X switched to mode Y every timestep anymore. I am not sure right now if this change happened before the 2018.10 release, I think perhaps not. In any case, this particular annoyance is dealt with.

**Comment by atgeirr** [comment]
Another point is that we should use the same terms as used in the deck format to describe the well control changes, i.e. ORAT, WRAT etc. rather than SURFACE_RATE. This will be made easier/possible by the changes discussed in #1699.

**Comment by GitPaean** [comment]
> When this happens then there is a MESSAGE to the console. And as long as there are not further control changes then there are no more MESSAGES.

Thanks for the information.  We probably should keep the `actual` control switch on the screen. 

> The impression I get from the Norne model, is that when Flow switches the control mode from SURFACE-RATE to BHP it issues a MESSAGE. However, at the next time step, it appears that still tries to operate the well at SURFACE-RATE, which it cannot, so it then changes control to BHP and issues a message. And I think his is why we are getting all the messages.

The reason for this is that, for Norne, as a history matching case, for each report step, there are some `WCON` keywords. When there are some new `WCON` keywords happen, the current design is that we try to begin with the control/target specified in the `WCON` keyword. The reason is that, between different report steps, the `WCON` keywords can be extremely flexible.  You can have any `WCON` keywords after another any `WCON` keywords.  To be safe, we decide to begin with the control mode specified in the new `WCON` keyword. Since the old control mode might not exist anymore or it is very dangerous to begin with that one. 

> For example, the control changes from ORAT to BHP, then the well switches to BHP control and the ORAT becomes a constraint.

It is correct in OPM-flow. But the solution procedure might have some back-and-forth iteration of the control mode to get finally converged.  There is some ideas how to improve the solution procedure. This might change in the future.

**Comment by GitPaean** [comment]
> Another point is that we should use the same terms as used in the deck format to describe the well control changes, i.e. ORAT, WRAT etc. rather than SURFACE_RATE. This will be made easier/possible by the changes discussed in #1699.

Yes.  Especially when you see switching from SURFACE_RATE to SURFACE_RATE.

**Comment by alfbr** [comment]
> > Yes, convergence information has been requested by users. What we have so far is in the PRT file does not exactly fit user requests. This we want to improve.
> 
> That's my point: I don't want to not provide that information, I only want it to go to the PRT and DBG files, not to the terminal.

My impression is the same as voiced by @OPMUSER , users want to see when well controls switch. Actually, at my end even more verbose messages on well mode changes has been asked for. More specifically, users in some cases want more information on why the well changed mode. I believe users would not like to look at the DBG file at all, but rather have everything they are interested in from that file in the PRT file. I think most users want well mode switching messages to STDOUT/LOG in addition to the PRT file, I may be wrong on this though.

**Comment by andlaus** [comment]
> You do not get repeated messages about well X switched to mode Y every timestep anymore. 

@atgeirr: I do still get this on norne:

```
Time step 1, stepsize 11 days.
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.900000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.800000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.700000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.600000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from SURFACE_RATE to BHP
    Oscillating behavior detected: Relaxation set to 0.500000
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Oscillating behavior detected: Relaxation set to 0.500000

Problem: Solver convergence failure - Iteration limit reached
Timestep chopped to 3.630000 days


Time step 1, stepsize 3.63 days.
    Switching control mode for well C-1H from BHP to SURFACE_RATE
    Switching control mode for well C-1H from SURFACE_RATE to BHP
Time step summary: well its =  2, newton its =  4, linearizations =  5 ( 0.722 sec), linear its =  53 ( 0.484 sec)

Time step 2, stepsize 7.37 days.
Time step summary: well its =  1, newton its =  3, linearizations =  4 ( 0.547 sec), linear its =  63 ( 0.554 sec)

Report step 172/247 at day 2177/3312, date = 23-Oct-2003
```

for cases with more wells this is bound to be worse.

**Comment by atgeirr** [comment]
@andlaus What you are seeing is back-and-forth switching within the same timestep, the behaviour that was fixed was more like: "step 7, well PROD switch to BHP; step 8, well PROD switch to BHP; step 9, well PROD switch to BHP....". Getting rid of the iteration-to-iteration switching we are seeing here is possible as @GitPaean suggested by only writing a message if it was switched at the end of the timestep.

The question is then: @OPMUSER, @alfbr, is the output above (showing control hunting within a timestep, eventually resulting in a timestep cut):
 - user-useful enough to retain in the stdout output
 - somewhat useful, but should be in PRT
 - not so useful, stick it in DBG.

In all cases, a report will be logged if, at the end of a timestep, control switched. The above question is about switching from iteration to iteration within a single step.

**Comment by andlaus** [comment]
@atgeirr: I don't normally care about this, and thus want a way to disable it. This information is not even useful as long as you don't know what iterations are affected. I'm fine if "Well XYZ swichted to control mode FOO" is printed once at the end of substeps, though.

**Comment by OPMUSER** [comment]
Oscillating behaviour is important because it can imply inconsistencies in the well data. For example the the VLP tables have crossing outflow curves or the VLP curves are extrapolated to unrealistic values. This will cause a performance hit and may result in inaccurate computations.

The other reason for the oscillating behaviour is that the well is close to dying, and small changes in parameters can switch the well on or off. In case there should be a message stating the fact.

**Comment by OPMUSER** [comment]
**_Re-Posted Due to Browser Issue_**

**Short answer:**
The question is then: @OPMUSER, @alfbr, is the output above (showing control hunting within a timestep, eventually resulting in a timestep cut):  **user-useful enough to retain in the stdout output**

**Long Answer:**
Concerning the usefulness of the change of well control MESSAGE, the commercial simulator gives this message: 
 ```
@--MESSAGE  AT TIME      365.0   DAYS    ( 1-JAN-2018):
 @           ******************************************                      
 @           * WELL PA2_W6   HAS CHANGED TO           *                      
 @           * CONTROL BY LIQUID RATE                 *                      
 @           * FROM TIME    365.00   DAYS  .          *                      
 @           ******************************************
```
Note the time and date so the engineers know when and what is happening. So may be Flow and give something like this:
 ```
AT TIME      365.0   DAYS    ( 1-JAN-2018): WELL PA2_W6  HAS `CHANGED` CONTROL FROM BHP TO LIQUID RATE                        
```
Oscillating behaviour is important because it can imply inconsistencies in the well data. For example the the VLP tables may have crossing outflow curves or the VLP curves are extrapolated to unrealistic values. This will cause a performance hit and may result in inaccurate computations. So this needs to be addressed by the engineers.

The other reason for the oscillating behaviour is that the well is close to dying, and small changes in solution parameters can switch the well on or off. In these cases there should be a message stating the fact. Below is an example for when a well is dying from the commercial simulator. Notice the classification of the messages (WARNING and MESSAGE).
```
 @--WARNING  AT TIME     1508.0   DAYS    (17-FEB-2021):
 @           THE RESERVOIR PRESSURE IS TOO LOW FOR WELL PA2_W5               
 @           TO PRODUCE AT THE REQUIRED BOTTOM HOLE PRESSURE                 

 @--MESSAGE  AT TIME     1508.0   DAYS    (17-FEB-2021):
 @           WELL PA2_W5   CANNOT OPERATE UNDER THE CURRENT                  
 @           RESERVOIR CONDITIONS IN NON-LINEAR ITERATION   0.               
 @           ITS IPR IS    Q(P)  =  A  -  BHP * B ,   WHERE:                 
 @           A =   2.3243E+03  3.2694E+03  9.2894E+02                        
 @           B =   4.6691E+00  6.5677E+00  1.8661E+00                        
 @           THE WELL IS STOPPED FOR THE ITERATION.                          
 @           LAST CONVERGED SOLUTION IS:                                     
 @           BHP       5.0000E+02                                            
 @           THP       0.0000E+00                                            
 @           RATE   OIL  :  3.8106E+01                                       
 @           RATE   WAT  :  4.7701E+01                                       
 @           RATE   GAS  :  1.4531E+01                                       

 @--MESSAGE  AT TIME     1508.0   DAYS    (17-FEB-2021):
 @           WELL PA2_W5   IS REVIVED FOR ITERATION   1 
```
So may be Flow can write something like this out.

```
AT TIME     1508.0   DAYS    (17-FEB-2021):  WELL PA2_W5 THE RESERVOIR PRESSURE IS TOO LOW FOR THE WELL TO PRODUCE AT THE REQUIRED BHP                 
                                          :  WELL PA2_W5   CANNOT OPERATE UNDER THE CURRENT RESERVOIR CONDITIONS IN NON-LINEAR ITERATION   0.   THE WELL IS STOPPED FOR THE ITERATION.                          
                                          :  WELL PA2_W5   IS REVIVED FOR ITERATION   1 
                                          :  WELL PA2_W5   CANNOT OPERATE UNDER THE CURRENT RESERVOIR CONDITIONS IN NON-LINEAR ITERATION   1.   THE WELL IS STOPPED FOR THE ITERATION.                          
                                          :  WELL PA2_W5   IS REVIVED FOR ITERATION   2

```
Or something similar.

All the above messages are sent to the PRT and LOG files and I think the terminal as well but I'm not 100% sure on that . Flow should do the same.

Note the commercial simulator has keyword _WVFPEXP Options to Overcome Certain Problems with THP Controlled Wells_, including making the VLP curve lookup explicitly. This means that VLP interpolation is from the previous time step (according to the manual).

Rather than add another command line parameter why don't we implement the `MESSAGES `keyword instead, so that developers and engineers can set the level and number of messages written out. This will give greater flexibility and hopefully keep everybody happy.

Sorry for the long post.

**Comment by OPMUSER** [comment]
I don't think so, it is written up in the manual but the keyword is classified as inactive.

**Comment by OPMUSER** [comment]
Note MESSAGES is a global keyword so it can be used in any section and throughout the SCHEDULE section. There are also six levels going from MESSAGE all the way to BUG. The keyword sets not only the print limits but also the stop limits.

**Comment by atgeirr** [comment]
> Note MESSAGES is a global keyword so it can be used in any section and throughout the SCHEDULE section. There are also six levels going from MESSAGE all the way to BUG. The keyword sets not only the print limits but also the stop limits.

I think our support of this is limited in that we do not consider stop limits, and we also do not allow reconfiguration during the SCHEDULE.

**Comment by atgeirr** [comment]
> I don't think so, it is written up in the manual but the keyword is classified as inactive.

I'll check.

**Comment by atgeirr** [comment]
I have now checked our MESSAGES support. It is as I thought: we support print but not stop limits, and only the initial state of MESSAGES is used. The relevant code (from FlowMainEbos.hpp) is:
```
const auto& msgLimits = schedule().getMessageLimits();
const std::map<int64_t, int> limits = {
    {Log::MessageType::Note, msgLimits.getCommentPrintLimit(0)},
    {Log::MessageType::Info, msgLimits.getMessagePrintLimit(0)},
    {Log::MessageType::Warning, msgLimits.getWarningPrintLimit(0)},
    {Log::MessageType::Error, msgLimits.getErrorPrintLimit(0)},
    {Log::MessageType::Problem, msgLimits.getProblemPrintLimit(0)},
    {Log::MessageType::Bug, msgLimits.getBugPrintLimit(0)}};
streamLog->setMessageLimiter(std::make_shared<MessageLimiter>(10, limits));
```
The `0` argument, for example in `getBugPrintLimit(0)`, is the report step. So in theory we could update the streamLog object to support MESSAGES dynamically. (The `10` in the MessageLimiter constructor is the limit for tagged messages.)

**Comment by atgeirr** [comment]
I do not think we should require exact match with Eclipse in terms of message categories, but we can try to make it close enough to be comfortable for engineers. Our logging system supports adding a prefix to indicate message category, in addition to color coding, but we have by default only enabled the prefix for warning, error, problem and bug messages today. It is easy to change this default, but the most basic information (the Info category) which is used for things like "Report step  0/247 at day 0/3312, date = 06-Nov-1997" should certainly not have a prefix. It is quite possible to make this behaviour controllable from the command line. We can add message categories too, if necessary.

One line per message is a bit strict for some of the longer messages, but better use of indentation is a good idea (there has been some minor cleanup of this before).

I think, if you are dissatisfied with any particular log message, stating a) where you want it to appear (stdout, PRT, DBG), b) if a colour or prefix is wanted (showing the category), c) if it should be limited (i.e. not appearing any more after some number of messages [default 10] about the same thing appears).

About the well messages in particular: we are right now making a better, more general system for ensuring such messages are dealt with properly when running in parallel. We may take the opportunity to revisit behaviour as well.

**Comment by OPMUSER** [comment]
@atgeirr When the maximum number of Info messages is reached one gets this:
```
    Switching control mode for well OP22 from SURFACE_RATE to BHP
    Switching control mode for well OP23 from SURFACE_RATE to BHP
    Switching control mode for well OP24 from SURFACE_RATE to BHP
Message limit reached for message category: Info
```
and nothing else is printed to the terminal. This is not correct as one should always get the time step information printed to the terminal and print file, that is:
```
Report step  0/121 at day 0/3652, date = 01-Jan-2016

Time step 0, stepsize 1 days.
```
In addition one should always get the final run summary at the end of the run, that is:
```
================    End of simulation     ===============

Total time (seconds):         736.939
Solver time (seconds):        704.363
 Assembly time (seconds):     205.746 (Failed: 9.24183; 4.49186%)
 Linear solve time (seconds): 465.479 (Failed: 22.3818; 4.80834%)
 Update time (seconds):       11.2274 (Failed: 0.616815; 5.49384%)
 Output write time (seconds): 38.985
Overall Well Iterations:      960 (Failed: 10; 1.04167%)
Overall Linearizations:       2007 (Failed: 84; 4.18535%)
Overall Newton Iterations:    1670 (Failed: 84; 5.02994%)
Overall Linear Iterations:    25913 (Failed: 1254; 4.83927%)
```

**Comment by OPMUSER** [comment]
I'm just running some OPM examples for a course I teach, so this is not a problem for me to switch back to the default value. However, may be:
`
Switching control mode for well OP22 from SURFACE_RATE to BHP`
Should no be at the INFO level. 
Whatever way you handle this  you must always print the time step information and the End of Simulation summary.

**Comment by atgeirr** [comment]
> Should no be at the INFO level.

Making it note() will put it only in PRT, is that better?

**Comment by OPMUSER** [comment]
@atgeirr 

> Should no be at the INFO level.

No, these type of messages should always be to the terminal as the user may wish to abort the job, as there may be something fundamentally wrong with the run.

Let's see what the next release brings before making any further changes.

One other comment the colour coding of the message is really nice when you run from the terminal. OPMRUN users the `tee` command to copy the terminal output to a *.LOG file similar to the commercial simulator, but this then has the "color codes" embedded in the output. Is there a command line option to switch this feature off, or is there an easy way to strip out these codes? 
```

Time step summary: well its =  1, newton its =  6, linearizations =  7 ( 0.712 sec), linear its = 152 ( 2.328 sec)
[34;1m
Warning: Keyword 'BASIC' is unhandled for output to file.[0m
[34;1m
Warning: Keyword 'CPU' is unhandled for output to file.[0m
[34;1m
Warning: Keyword 'NEWTON' is unhandled for output to file.[0m
[34;1m
Warning: Keyword 'RESTART' is unhandled for output to file.[0m
[34;1m
Warning: Keyword 'SUMMARY' is unhandled for output to file.[0m
[34;1m
Warning: Message limit reached for message tag: Unhandled output keyword[0m

```

**Comment by atgeirr** [comment]
> Is there a command line option to switch this feature off, or is there an easy way to strip out these codes?

No there is no such command line option, but it would not be a lot of work to implement if required? On the other hand, the output in PRT and DBG (which includes the stdout output) is not color coded. I do not know a simple way to strip them. Perhaps the simplest way is to always write a LOG file with the same messages as those written to the terminal, except not color coded?

**Comment by OPMUSER** [comment]
> Perhaps the simplest way is to always write a LOG file with the same messages as those written to the terminal, except not color coded?

That will work for me.

**Comment by alfbr** [comment]
> Perhaps the simplest way is to always write a LOG file with the same messages as those written to the terminal

Please let us look at alternatives first. Unconditionally writing out LOG file is unexpected behaviour with unexpected consequences over here.

**Comment by atgeirr** [comment]
> Please let us look at alternatives first. Unconditionally writing out LOG file is unexpected behaviour with unexpected consequences over here.

Ok, I will not do anything about this right now (doing it unconditionally would make it a small and simple code change though: ~3 lines).

**Comment by atgeirr** [comment]
> I do not know a simple way to strip them.

I do now, if you are still interested:
```
perl -pe 's/\x1b\[[0-9;]*[mG]//g'
```
works cross-platform and is not too unreadable. See https://superuser.com/questions/380772/removing-ansi-color-codes-from-text-stream for info.

**Comment by OPMUSER** [comment]
Thanks @atgeirr I'll see how I can incorporate this automatically in OPMRUN.
